File: 633.sum-of-square-numbers.py
#
# @lc app=leetcode id=633 lang=python
#
# [633] Sum of Square Numbers
#

# @lc code=start
class Solution(object):
    def judgeSquareSum(self, c):
        """
        :type c: int
        :rtype: bool
        """
        start = 0
        end = int(c ** 0.5)
        while start <= end:
            cur = start*start + end*end
            if cur < c:
                start += 1
            elif cur > c:
                end -= 1
            else:
                return True
        return False
# @lc code=end

File: 633.sum-of-square-numbers.cpp
/*
 * @lc app=leetcode id=633 lang=cpp
 *
 * [633] Sum of Square Numbers
 */
#include <math.h>

// @lc code=start
class Solution {
public:
    bool judgeSquareSum(int c) {
        if (c<0)
            return false;
        long int left = 0, right = sqrt(c);
        while (left <= right) {
            long int cur = left * left + right * right;
            if (cur <c)
                left++;
            else if (cur >c)
                right--;
            else
                return true;
        }
        return false;
    }
};
// @lc code=end

